pkg/tunnel/relay.go

```diff
@@ -0,0 +1,355 @@
+		http.Error(w, "Failed to update virtual network keys", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusCreated)
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		http.Error(w, "Failed to encode response", http.StatusInternalServerError)
+		return
+	}
+}
+
+func (r *Relay) handleDisconnect(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	var request api.Request
+	if err := json.NewDecoder(req.Body).Decode(&request); err != nil && !errors.Is(err, io.EOF) {
+		http.Error(w, "Invalid request", http.StatusBadRequest)
+		return
+	}
+
+	r.mu.Lock()
+	onDisconnect := r.onDisconnect
+	r.mu.Unlock()
+
+	if err := onDisconnect(req.Context(), request.Agent, request.ID); err != nil {
+		slog.Error("onDisconnect callback failed", slog.Any("error", err))
+		http.Error(w, "Failed to handle disconnection", http.StatusInternalServerError)
+		return
+	}
+
+	r.conns.Del(request.ID)
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (r *Relay) handleUpdateKeys(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	var request api.Request
+	if err := json.NewDecoder(req.Body).Decode(&request); err != nil && !errors.Is(err, io.EOF) {
+		http.Error(w, "Invalid request", http.StatusBadRequest)
+		return
+	}
+
+	conn, ok := r.conns.Get(request.ID)
+	if !ok {
+		http.Error(w, "Connection not found", http.StatusNotFound)
+		return
+	}
+
+	vni := conn.VNI()
+	if vni == nil {
+		slog.Warn("Connection has no VNI assigned", slog.String("connID", request.ID))
+		http.Error(w, "Connection is not ready", http.StatusBadRequest)
+		return
+	}
+
+	sendKey, err := randomKey()
+	if err != nil {
+		http.Error(w, "Failed to generate key", http.StatusInternalServerError)
+		return
+	}
+
+	recvKey, err := randomKey()
+	if err != nil {
+		http.Error(w, "Failed to generate key", http.StatusInternalServerError)
+		return
+	}
+
+	keys := api.Keys{
+		Epoch:     conn.IncrementKeyEpoch(),
+		Send:      sendKey,
+		Recv:      recvKey,
+		ExpiresAt: time.Now().Add(keyLifespan),
+	}
+
+	if err := r.handler.UpdateVirtualNetworkKeys(*vni, keys.Epoch,
+		keys.Send, keys.Recv, keys.ExpiresAt); err != nil {
+		http.Error(w, "Failed to update virtual network keys", http.StatusInternalServerError)
+		return
+	}
+
+	resp := api.UpdateKeysResponse{
+		Keys: keys,
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		http.Error(w, "Failed to encode response", http.StatusInternalServerError)
+		return
+	}
+}
+
+func (r *Relay) withAuth(next httprouter.Handle) httprouter.Handle {
+	return func(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+		authHeader := req.Header.Get("Authorization")
+		const prefix = "Bearer "
+		if authHeader == "" || len(authHeader) <= len(prefix) || authHeader[:len(prefix)] != prefix {
+			http.Error(w, "Unauthorized", http.StatusUnauthorized)
```

Comment:
Surprisingly this doesn't cause transport teardown so if client holds the quic connection and continues to send heartbeats the server never disconnects. We had a case in prod where we accumulated thousands of dormant QUIC streams because the client didn't cancel properly. See my hacks in https://github.com/apoxy-dev/apoxy/commit/682d735ea65cfc0183a734f91d16ee8ecf24cbbb to address this. Lmk if you have better ideas